File: fsharp/protein-translation/_iterations/1/ProteinTranslation.fs
module ProteinTranslation

let private protein codon =
    match codon with
    | [| 'A'; 'U'; 'G' |] -> "Methionine"
    | [| 'U'; 'U'; 'U' |] | [| 'U'; 'U'; 'C' |] -> "Phenylalanine"
    | [| 'U'; 'U'; 'A' |] | [| 'U'; 'U'; 'G' |] -> "Leucine"
    | [| 'U'; 'C'; 'U' |] | [| 'U'; 'C'; 'C' |]
    | [| 'U'; 'C'; 'A' |] | [| 'U'; 'C'; 'G' |] -> "Serine"
    | [| 'U'; 'A'; 'U' |] | [| 'U'; 'A'; 'C' |] -> "Tyrosine"
    | [| 'U'; 'G'; 'U' |] | [| 'U'; 'G'; 'C' |] -> "Cysteine"
    | [| 'U'; 'G'; 'G' |] -> "Tryptophan"
    | [| 'U'; 'A'; 'A' |] | [| 'U'; 'A'; 'G' |] | [| 'U'; 'G'; 'A' |] -> "STOP"
    | _ -> failwith $"Invalid codon: {codon}"

let proteins (rna: string): string list =
    rna
    |> Seq.chunkBySize 3
    |> Seq.map protein
    |> Seq.takeWhile ((<>) "STOP")
    |> Seq.toList

File: clojure/protein-translation/_iterations/1/src/protein_translation.clj
(ns protein-translation)

(def ^:private proteins-to-codons
  {"Methionine"    ["AUG"]
   "Phenylalanine" ["UUU" "UUC"]
   "Leucine"       ["UUA" "UUG"]
   "Serine"        ["UCU" "UCC" "UCA" "UCG"]
   "Tyrosine"      ["UAU" "UAC"]
   "Cysteine"      ["UGU" "UGC"]
   "Tryptophan"    ["UGG"]
   "STOP"          ["UAA" "UAG" "UGA"]})

(def ^:private codons-to-protein
  (into {} (for [[protein codons] proteins-to-codons
                 codon codons]
             [codon protein])))

(defn translate-codon [codon]
  (codons-to-protein codon))

(defn translate-rna [rna]
  (loop [[nucleotide1 nucleotide2 nucleotide3 & other-nucleotides] rna
         proteins []]
    (let [protein (translate-codon (str nucleotide1 nucleotide2 nucleotide3))]
      (if (or (nil? protein) (= protein "STOP"))
        proteins
        (recur other-nucleotides (conj proteins protein))))))
